File: androidApp/src/main/java/org/idos/app/ui/screens/credentials/CredentialDetailViewModel.kt
package org.idos.app.ui.screens.credentials

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.launch
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import org.idos.app.data.StorageManager
import org.idos.app.data.model.CredentialDetail
import org.idos.app.data.repository.CredentialsRepository
import org.idos.app.navigation.NavRoute
import org.idos.app.navigation.NavigationCommand
import org.idos.app.navigation.NavigationManager
import org.idos.app.ui.screens.base.BaseViewModel
import org.idos.app.ui.screens.settings.EnclaveUiStatus
import org.idos.enclave.DecryptFailure
import org.idos.enclave.Enclave
import org.idos.enclave.EnclaveError
import org.idos.enclave.EnclaveKeyType
import org.idos.enclave.EnclaveOrchestrator
import org.idos.enclave.EnclaveSessionConfig
import org.idos.enclave.EnclaveState
import org.idos.enclave.ExpirationType
import org.idos.kwil.serialization.toByteArray
import org.idos.kwil.types.Base64String
import org.idos.kwil.types.UuidString
import timber.log.Timber
import kotlin.time.Duration

sealed class CredentialDetailState {
    data object Loading : CredentialDetailState()

    data class Loaded(
        val credential: CredentialDetail,
        val decryptedContent: JsonElement? = null, // null = still encrypted
        val enabled: Boolean = true,
    ) : CredentialDetailState()

    data class Error(
        val message: String,
    ) : CredentialDetailState()
}

sealed class CredentialDetailEvent {
    data object LoadCredential : CredentialDetailEvent()

    data object DecryptCredential : CredentialDetailEvent()

    data object Retry : CredentialDetailEvent()

    data object Back : CredentialDetailEvent()

    // Enclave events
    data class UnlockEnclave(
        val password: String?,
        val expiration: EnclaveSessionConfig,
    ) : CredentialDetailEvent()

    data object LockEnclave : CredentialDetailEvent()

    data object DismissEnclave : CredentialDetailEvent()

    data object RetryDecrypt : CredentialDetailEvent()
}

sealed class EnclaveUiState(
    open val type: EnclaveKeyType? = null,
) {
    data object Hidden : EnclaveUiState(type = null)

    data class Unlocking(
        override val type: EnclaveKeyType,
    ) : EnclaveUiState(type)

    data class RequiresUnlock(
        override val type: EnclaveKeyType,
    ) : EnclaveUiState(type)

    data class UnlockError(
        override val type: EnclaveKeyType,
        val message: String,
        val canRetry: Boolean = true,
    ) : EnclaveUiState(type)
}

class CredentialDetailViewModel(
    private val credentialsRepository: CredentialsRepository,
    private val navigationManager: NavigationManager,
    private val storageManager: StorageManager,
    private val orchestrator: EnclaveOrchestrator,
    savedStateHandle: SavedStateHandle,
) : BaseViewModel<CredentialDetailState, CredentialDetailEvent>() {
    val credentialId: UuidString =
        requireNotNull(savedStateHandle.get<String>(NavRoute.CredentialDetail.CREDENTIAL_ID_ARG))

    // Track enclave UI state (dialog, errors)
    private val _enclaveUiState = MutableStateFlow<EnclaveUiState>(EnclaveUiState.Hidden)
    val enclaveUiState = _enclaveUiState

    init {
        viewModelScope.launch {
            orchestrator.checkStatus()
        }
        loadCredential()
        observeEnclaveState()
    }

    private fun observeEnclaveState() {
        viewModelScope.launch {
            orchestrator.state.collect { state ->
                when (state) {
                    is EnclaveState.Unlocked -> {
                        Timber.i("Enclave unlocked successfully")
                        _enclaveUiState.value = EnclaveUiState.Hidden
                        decryptLoadedCredential()
                    }

                    is EnclaveState.Unlocking -> {
                        _enclaveUiState.value = EnclaveUiState.Unlocking(
                            type = orchestrator.getEnclaveType()
                        )
                    }

                    is EnclaveState.Locked -> {
                        // Don't automatically show dialog - wait for decrypt attempt
                    }

                    is EnclaveState.NotAvailable -> {
                        // Noop, credential detail will pass NA state to UI; enclave is not mandatory
                    }
                }
            }
        }
    }

    private fun loadCredential() {
        viewModelScope.launch {
            try {
                updateState { CredentialDetailState.Loading }

                credentialsRepository
                    .getCredential(credentialId)
                    .collect { detail ->
                        val enabled = orchestrator.state.value !is EnclaveState.NotAvailable
                        updateState { CredentialDetailState.Loaded(detail, decryptedContent = null, enabled) }
                        // Try to decrypt if enclave is unlocked
                        if (orchestrator.state.value is EnclaveState.Unlocked) {
                            decryptLoadedCredential()
                        }
                    }
            } catch (e: Exception) {
                Timber.e(e, "Failed to load credential")
                updateState { CredentialDetailState.Error("Failed to load credential: ${e.message}") }
            }
        }
    }

    private fun decryptLoadedCredential() {
        val currentState = state.value
        if (currentState !is CredentialDetailState.Loaded) return
        if (currentState.decryptedContent != null) return // Already decrypted

        viewModelScope.launch {
            try {
                orchestrator
                    .withEnclave { enclave ->
                        val decryptedString = decryptCredential(currentState.credential, enclave)
                        val json = Json.parseToJsonElement(decryptedString)
                        updateState {
                            (this as CredentialDetailState.Loaded).copy(decryptedContent = json)
                        }
                    }
            } catch (e: Exception) {
                when (e) {
                    is EnclaveError.NoKey, is EnclaveError.KeyExpired -> {
                        Timber.i("Enclave locked, prompting for password")
                        _enclaveUiState.value = EnclaveUiState.RequiresUnlock(orchestrator.getEnclaveType())
                    }

                    is EnclaveError.DecryptionFailed -> {
                        val message =
                            when {
                                e.reason == DecryptFailure.WrongPassword -> "Wrong password - key cannot decrypt this data"
                                else -> "Decryption failed: ${e.message}"
                            }
                        _enclaveUiState.value = EnclaveUiState.UnlockError(
                            type = orchestrator.getEnclaveType(),
                            message = message,
                            canRetry = true
                        )
                    }

                    else -> {
                        Timber.e(e, "Decryption error")
                        updateState { CredentialDetailState.Error("Failed to decrypt: ${e.message}") }
                    }
                }
            }
        }
    }

    private suspend fun decryptCredential(
        data: CredentialDetail,
        enclave: Enclave,
    ): String {
        val content = Base64String(data.content).toByteArray()
        val pubkey = Base64String(data.encryptorPublicKey).toByteArray()
        return enclave.decrypt(content, pubkey).decodeToString()
    }

    private fun unlockEnclave(event: CredentialDetailEvent.UnlockEnclave) {
        viewModelScope.launch {
            try {
                val user = storageManager.getStoredUser() ?: error("no user stored")
                orchestrator.unlock(user.id, event.expiration, event.password)
            } catch (e: EnclaveError) {
                Timber.e(e, "Failed to unlock enclave")
                _enclaveUiState.value =
                    EnclaveUiState.UnlockError(
                        type = orchestrator.getEnclaveType(),
                        message = e.message ?: "Failed to unlock enclave",
                        canRetry = false,
                    )
            }
        }
    }

    private fun lockEnclave() {
        viewModelScope.launch {
            orchestrator.lock()
            updateState {
                if (this is CredentialDetailState.Loaded) {
                    copy(decryptedContent = null)
                } else {
                    this
                }
            }
        }
    }

    private fun dismissEnclave() {
        _enclaveUiState.value = EnclaveUiState.Hidden
//        viewModelScope.launch {
//            navigationManager.navigate(NavigationCommand.NavigateUp)
//        }
    }

    private fun retryDecrypt() {
        _enclaveUiState.value = EnclaveUiState.Hidden
        decryptLoadedCredential()
    }

    override fun initialState(): CredentialDetailState = CredentialDetailState.Loading

    override fun onEvent(event: CredentialDetailEvent) {
        when (event) {
            is CredentialDetailEvent.LoadCredential -> loadCredential()
            is CredentialDetailEvent.DecryptCredential -> {
                // Trigger decrypt flow
                when (orchestrator.state.value) {
                    is EnclaveState.Unlocked -> decryptLoadedCredential()
                    is EnclaveState.Locked -> _enclaveUiState.value = EnclaveUiState.RequiresUnlock(orchestrator.getEnclaveType())
                    else -> {}
                }
            }

            is CredentialDetailEvent.Retry -> loadCredential()
            is CredentialDetailEvent.Back -> {
                viewModelScope.launch {
                    navigationManager.navigate(NavigationCommand.NavigateUp)
                }
            }

            is CredentialDetailEvent.UnlockEnclave -> unlockEnclave(event)
            is CredentialDetailEvent.LockEnclave -> lockEnclave()
            is CredentialDetailEvent.DismissEnclave -> dismissEnclave()
            is CredentialDetailEvent.RetryDecrypt -> retryDecrypt()
        }
    }
}

File: iosApp/iosApp/Features/Credentials/CredentialDetailViewModel.swift
import Foundation
import UIKit
import Combine
import OSLog
import idos_sdk

// MARK: - Enclave UI State

/// Enclave dialog/UI state (matches Android EnclaveUiState)
enum EnclaveUiState: Equatable {
    case hidden
    case requiresUnlock(type: EnclaveKeyType)
    case unlocking(type: EnclaveKeyType)
    case unlockError(type: EnclaveKeyType, message: String, canRetry: Bool)

    var type: EnclaveKeyType? {
        switch self {
        case .hidden:
            return nil
        case .requiresUnlock(let type), .unlocking(let type), .unlockError(let type, _, _):
            return type
        }
    }

    // Manual Equatable conformance because EnclaveKeyType is from Kotlin
    static func == (lhs: EnclaveUiState, rhs: EnclaveUiState) -> Bool {
        switch (lhs, rhs) {
        case (.hidden, .hidden):
            return true
        case (.requiresUnlock(let lType), .requiresUnlock(let rType)),
             (.unlocking(let lType), .unlocking(let rType)):
            return lType == rType
        case (.unlockError(let lType, let lMsg, let lRetry), .unlockError(let rType, let rMsg, let rRetry)):
            return lType == rType && lMsg == rMsg && lRetry == rRetry
        default:
            return false
        }
    }
}

// MARK: - Credential Detail State

/// Credential Detail state
struct CredentialDetailState {
    var credential: CredentialDetail?
    var decryptedContent: String? = nil  // nil = still encrypted
    var isLoading: Bool = false
    var isDecrypting: Bool = false
    var error: String? = nil
}

// MARK: - Credential Detail Events

/// Credential Detail events
enum CredentialDetailEvent {
    case loadCredential
    case decryptCredential
    case clearError

    // Enclave events
    case unlockEnclave(password: String?, config: EnclaveSessionConfig)
    case lockEnclave
    case dismissEnclave
}

// MARK: - ViewModel

/// CredentialDetailViewModel with direct orchestrator usage (matches Android pattern)
class CredentialDetailViewModel: BaseViewModel<CredentialDetailState, CredentialDetailEvent> {
    private let credentialId: String
    private let navigationCoordinator: NavigationCoordinator
    private let credentialsRepository: CredentialsRepositoryProtocol
    private let userRepository: UserRepositoryProtocol
    private let orchestrator: EnclaveOrchestrator

    @Published var enclaveUiState: EnclaveUiState = .hidden

    init(
        credentialId: String,
        orchestrator: EnclaveOrchestrator,
        credentialsRepository: CredentialsRepositoryProtocol,
        userRepository: UserRepositoryProtocol,
        navigationCoordinator: NavigationCoordinator
    ) {
        Logger.viewModel.debug("CredentialDetailViewModel: Initializing for credentialId: \(credentialId)")
        self.credentialId = credentialId
        self.navigationCoordinator = navigationCoordinator
        self.credentialsRepository = credentialsRepository
        self.userRepository = userRepository
        self.orchestrator = orchestrator

        super.init(initialState: CredentialDetailState())

        loadCredential()
        observeEnclaveState()
        
        // Check status on init
        Task {
            do {
                try await orchestrator.checkStatus()
            } catch {}
        }
    }

    override func onEvent(_ event: CredentialDetailEvent) {
        Logger.viewModel.debug("CredentialDetailViewModel: on event: \(String(describing: event))")
        switch event {
        case .loadCredential:
            loadCredential()
        case .decryptCredential:
            triggerDecrypt()
        case .clearError:
            state.error = nil
        case .unlockEnclave(let password, let config):
            unlockEnclave(password: password, config: config)
        case .lockEnclave:
            lockEnclave()
        case .dismissEnclave:
            dismissEnclave()
        }
    }

    // MARK: - Enclave State Observer

    private func observeEnclaveState() {
        Task { @MainActor in
            for await enclaveState: EnclaveState in orchestrator.state {
                Logger.enclave.debug("CredentialDetailViewModel: Observing enclave state: \(String(describing: type(of: enclaveState)))")
                switch enclaveState {
                case is EnclaveState.Locked:
                    // Don't automatically show dialog - wait for decrypt attempt
                    break

                case is EnclaveState.Unlocking:
                    self.enclaveUiState = .unlocking(type: orchestrator.getEnclaveType())

                case let unlocked as EnclaveState.Unlocked:
                    self.enclaveUiState = .hidden
                    // Auto-decrypt when unlocked
                    guard let credential = state.credential else { return }
                    await decryptLoadedCredential(enclave: unlocked.enclave, credential: credential)

                case is EnclaveState.NotAvailable:
                    // Enclave not available, don't show dialog
                    break

                default:
                    break
                }
            }
        }
    }

    // MARK: - Load Credential

    private func loadCredential() {
        Logger.viewModel.debug("CredentialDetailViewModel: Starting loadCredential for id: \(self.credentialId)")
        Task { @MainActor [weak self] in
            guard let self = self else { return }

            self.state.isLoading = true
            self.state.error = nil
            
            do {
                let detail = try await credentialsRepository.getCredential(id: credentialId)

                let currentState = orchestrator.state.value
                if currentState is EnclaveState.Unlocked {
                    let unlocked = currentState as! EnclaveState.Unlocked
                    await decryptLoadedCredential(enclave: unlocked.enclave, credential: detail)
                }
                
                self.state.credential = detail
                self.state.isLoading = false
            } catch {
                Logger.viewModel.error("CredentialDetailViewModel: Load failed - \(error.localizedDescription)")
                self.state.error = "Failed to load credential: \(error.localizedDescription)"
                self.state.isLoading = false
            }
        }
    }

    // MARK: - Decrypt

    private func triggerDecrypt() {
        guard let credential = state.credential else { return }
        // Check if enclave is unlocked
        self.state.isDecrypting = true
        if orchestrator.state.value is EnclaveState.Unlocked {
            let unlocked = orchestrator.state.value as! EnclaveState.Unlocked
            Task {
                await decryptLoadedCredential(enclave: unlocked.enclave, credential: credential)
            }
        } else if orchestrator.state.value is EnclaveState.Locked {
            // Show unlock dialog
            enclaveUiState = .requiresUnlock(type: orchestrator.getEnclaveType())
        } else {
            // Enclave not available
            self.state.error = "Encryption key not available"
            self.state.isDecrypting = false
        }
    }

    private func decryptLoadedCredential(enclave: Enclave, credential: CredentialDetail) async {
        guard state.decryptedContent == nil else { return } // Already decrypted

        Logger.enclave.debug("CredentialDetailViewModel: Starting decryption")

        let encryptedData = Data(base64Encoded: credential.content) ?? Data()
        let senderPublicKey = Data(base64Encoded: credential.encryptorPublicKey) ?? Data()

        do {
            let decryptedData = try await enclave.decrypt(
                message: encryptedData.toKotlinByteArray(),
                senderPublicKey: senderPublicKey.toKotlinByteArray()
            )
            
            if let decryptedContent = String(data: decryptedData.toNSData(), encoding: .utf8) {
                await MainActor.run {
                    self.state.decryptedContent = decryptedContent
                    self.state.isDecrypting = false
                    Logger.enclave.info("CredentialDetailViewModel: Decryption successful")
                }
            } else {
                await MainActor.run {
                    self.state.error = "Failed to decode decrypted content"
                }
            }
        } catch {
            if let enclaveError = error.asEnclaveError() {
                await MainActor.run {
                    switch onEnum(of: enclaveError) {
                    case .noKey, .keyExpired:
                        Logger.enclave.notice("CredentialDetailViewModel: No key found, prompting for unlock")
                        self.enclaveUiState = .requiresUnlock(type: orchestrator.getEnclaveType())

                    case .decryptionFailed(let reason):
                        Logger.enclave.error("CredentialDetailViewModel: Decryption failed - \(reason)")
                        let message = reason.description()
                        self.enclaveUiState = .unlockError(
                            type: orchestrator.getEnclaveType(),
                            message: message,
                            canRetry: true
                        )
                        self.state.isDecrypting = false

                    default:
                        Logger.enclave.error("CredentialDetailViewModel: Decryption failed with error: \(enclaveError.message ?? "unknown")")
                        self.state.error = enclaveError.message
                        self.state.isDecrypting = false
                    }
                }
            } else {
                await MainActor.run {
                    self.state.error = "Decryption error: \(error.localizedDescription)"
                    self.state.isDecrypting = false
                }
            }
        }
    }

    // MARK: - Enclave Operations

    private func unlockEnclave(password: String?, config: EnclaveSessionConfig) {
        guard let user = userRepository.getStoredUser() else {
            state.error = "No user ID found"
            return
        }

        Task { @MainActor [weak self] in
            guard let self = self else { return }

            do {
                try await orchestrator.unlock(
                    userId: user.id,
                    sessionConfig: config,
                    password: password
                )
                Logger.enclave.info("CredentialDetailViewModel: Enclave unlocked successfully")
                // State observer will trigger decrypt
            } catch {
                Logger.enclave.error("CredentialDetailViewModel: Unlock failed - \(error.localizedDescription)")
                if let enclaveError = error.asEnclaveError() {
                    self.enclaveUiState = .unlockError(
                        type: orchestrator.getEnclaveType(),
                        message: enclaveError.message ?? "Failed to unlock enclave",
                        canRetry: false
                    )
                } else {
                    self.enclaveUiState = .unlockError(
                        type: orchestrator.getEnclaveType(),
                        message: error.localizedDescription,
                        canRetry: false
                    )
                }
                self.state.isDecrypting = false
            }
        }
    }

    private func lockEnclave() {
        Task { @MainActor in
            do {
                _ = try await orchestrator.lock()
                Logger.enclave.info("CredentialDetailViewModel: Enclave locked successfully")
            } catch {
                // noop
            }
            self.state.decryptedContent = nil
            self.state.error = nil
            self.state.isDecrypting = false
        }
    }

    private func dismissEnclave() {
        enclaveUiState = .hidden
        self.state.isDecrypting = false
    }
}
